code/5380-marko-spasovski/script.js

```diff
@@ -1,3 +1,392 @@
+    }else if(key.getAttribute('data-sort') === 'companion'){
+        if(isAscending){
+            const filteredData = coppiedData.sort((a, b) => getCompanion(a).localeCompare(getCompanion(b)))
+            isAscending = false;
+            key.classList.add("descending");
+            displayData(filteredData);
+        }else{
+            const filteredData = coppiedData.sort((a, b) =>  getCompanion(b).localeCompare(getCompanion(a)))
+            isAscending = true;
+            key.classList.add("ascending");
+            displayData(filteredData);
+        }
+    }else if(key.getAttribute('data-sort') === 'cast'){
+        if(isAscending){
+            const filteredData = coppiedData.sort((a, b) => a.cast.length - b.cast.length)
+            isAscending = false;
+            key.classList.add("descending");
+            displayData(filteredData);
+        }else{
+            const filteredData = coppiedData.sort((a, b) => b.cast.length - a.cast.length)
+            isAscending = true;
+            key.classList.add("ascending");
+            displayData(filteredData);
+        }
+    }else if(key.getAttribute('data-sort') === 'plot'){
+        if(isAscending){
+            const filteredData = coppiedData.sort((a, b) => a.plot.localeCompare(b.plot))
+            isAscending = false;
+            key.classList.add("descending");
+            displayData(filteredData);
+        }else{
+            const filteredData = coppiedData.sort((a, b) => b.plot.localeCompare(a.plot))
+            isAscending = true;
+            key.classList.add("ascending");
+            displayData(filteredData);
+        }
+    }
+}  
+const getHeaderDiv = () =>{
+    const div = document.createElement("div");
+
+    div.innerHTML = `
+            <div class="episodes-header">
+                <div class="header-cell" data-sort="rank" >Rank</div>
+                <div class="header-cell" data-sort="name">Name</div>
+                <div class="header-cell" data-sort="series">Series</div>
+                <div class="header-cell" data-sort="era">Era</div>
+                <div class="header-cell" data-sort="broadcast">Broadcast Year</div>
+                <div class="header-cell" data-sort="director">Director</div>
+                <div class="header-cell" data-sort="writer">Writer</div>
+                <div class="header-cell" data-sort="doctor">Doctor</div>
+                <div class="header-cell" data-sort="companion">Companion</div>
+                <div class="header-cell" data-sort="cast">Cast</div>
+                <div class="header-cell" data-sort="plot">Plot</div>
+            </div>
+    `
+
+    return div;
+}
+const getAllHeaders = () =>{
+    const headers = document.getElementsByClassName('header-cell');
+    
+    return [...headers];  //this may be an error
+}
+const fetchData = async() =>{
+    url = "https://raw.githubusercontent.com/sweko/internet-programming-a98db973kwl8xp1lz94kjf0bma5pez8c/refs/heads/main/data/doctor-who-episodes.json";
```

Comment:
Proper loading of the data (the `url` parameter should have a `const` before it)